devito/types/dense.py

```diff
@@ -1077,36 +1062,73 @@ def _eval_at(self, func):
             return self.subs(mapper)
         return self
 
+    @classmethod
+    def __staggered_setup__(cls, dimensions, staggered=None, **kwargs):
+        """
+        Setup staggering-related metadata. This method assigns:
+
+            * 0 to non-staggered dimensions;
+            * 1 to staggered dimensions.
+        """
+        if not staggered:
+            processed = ()
+        elif staggered is CELL:
+            processed = (sympy.S.One,)*len(dimensions)
+        elif staggered is NODE:
+            processed = (sympy.S.Zero,)*len(dimensions)
+        elif all(is_integer(s) for s in as_tuple(staggered)):
+            # Staggering is already a tuple likely from rebuild
+            assert len(staggered) == len(dimensions)
+            processed = staggered
+        else:
+            processed = []
+            for d in dimensions:
+                if d in as_tuple(staggered):
+                    processed.append(sympy.S.One)
+                elif -d in as_tuple(staggered):
+                    processed.append(sympy.S.NegativeOne)
+                else:
+                    processed.append(sympy.S.Zero)
+        return tuple(processed)
+
     @classmethod
     def __indices_setup__(cls, *args, **kwargs):
         grid = kwargs.get('grid')
         dimensions = kwargs.get('dimensions')
+        staggered = kwargs.get('staggered')
+
         if grid is None:
             if dimensions is None:
                 raise TypeError("Need either `grid` or `dimensions`")
         elif dimensions is None:
             dimensions = grid.dimensions
 
+        staggered = cls.__staggered_setup__(dimensions, staggered=staggered)
         if args:
             assert len(args) == len(dimensions)
-            return tuple(dimensions), tuple(args)
-
-        # Staggered indices
-        staggered = kwargs.get("staggered", None)
-        if staggered in [CELL, NODE]:
-            staggered_indices = dimensions
+            staggered_indices = tuple(args)
         else:
```

Comment:
ultra-nitpicking , I'd just go with `elif staggered: ...` and then `else` (ie denesting everything)